redash/query_runner/mongodb.py

```diff
@@ -102,14 +102,55 @@ def _get_column_by_name(self, columns, column_name):
 
         return None
 
-
-    def run_query(self, query):
+    def _get_db(self):
         if self.is_replica_set:
             db_connection = pymongo.MongoReplicaSetClient(self.configuration["connectionString"], replicaSet=self.configuration["replicaSetName"])
         else:
             db_connection = pymongo.MongoClient(self.configuration["connectionString"])
 
-        db = db_connection[self.db_name]
+        return db_connection[self.db_name]
+
+    def _merge_property_names(self, columns, document):
+        for property in document:
+              if property not in columns:
+                  columns.append(property)
+
+    def _get_collection_fields(self, db, collection_name):
+        # Since MongoDB is a document based database and each document doesn't have 
+        # to have the same fields as another documet in the collection its a bit hard to
+        # show these attributes as fields in the schema.
+        #
+        # For now, the logic is to take the first and last documents (last is determined
+        # by the Natural Order (http://www.mongodb.org/display/DOCS/Sorting+and+Natural+Order)
+        # as we don't know the correct order. In most single server installations it would be 
+        # find. In replicaset when reading from non master it might not return the really last
+        # document written.
+        first_document = None
+        last_document = None
+
+        for d in db[collection_name].find().sort([("$natural", 1)]).limit(1):
+            first_document = d
+
+        for d in db[collection_name].find().sort([("$natural", -1)]).limit(1):
+            last_document = d
+
+        columns = []
+        if first_document: self._merge_property_names(columns, first_document)
+        if last_document: self._merge_property_names(columns, last_document)
```

Comment:
:-)

Anyhow the reason for the merge  methods is that documents can contain sub
dictionary and while I didn't implement it yet I want to show subdocuments
in the from of collection.fieldname (similar to what Bigquery does to json)

On Sun, Dec 20, 2015, 14:38 Arik Fraimovich <notifications@github.com>
wrote:

> In redash/query_runner/mongodb.py
> <https://github.com/getredash/redash/pull/714#discussion_r48102960>:
>
> > +        # by the Natural Order (http://www.mongodb.org/display/DOCS/Sorting+and+Natural+Order)
> > +        # as we don't know the correct order. In most single server installations it would be
> > +        # find. In replicaset when reading from non master it might not return the really last
> > +        # document written.
> > +        first_document = None
> > +        last_document = None
> > +
> > +        for d in db[collection_name].find().sort([("$natural", 1)]).limit(1):
> > +            first_document = d
> > +
> > +        for d in db[collection_name].find().sort([("$natural", -1)]).limit(1):
> > +            last_document = d
> > +
> > +        columns = []
> > +        if first_document: self._merge_property_names(columns, first_document)
> > +        if last_document: self._merge_property_names(columns, last_document)
>
> I'm not sure the above qualifies as "syntax trickery" :-)
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/getredash/redash/pull/714/files#r48102960>.
>